### ISSUE TITLE
Oops, the program crashed with one of the test cases provided. 

### ISSUE THREAD

**Issue by cnitlrt** [AFLplusplus/AFLplusplus#1733]
Hello, I have encountered an error, I tried to solve it but failed
### environment
```
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04 LTS
Release:	20.04
Codename:	focal
```
### source
afl-gcc test.c -o 1 -fsanitize=address -g -static-libasan
```
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(int argc, const char *argv[]) {
    char *s = (char*)malloc(100);
    strcpy(s, "Hello world!");
    printf("string is: %s\n", s);
    return 0;
}
```
After I set the environment variable, running the file directly will not report an error. The following is the environment variable I set
ASAN_OPTIONS="disable_coredump=0:unmap_shadow_on_exit=1:abort_on_error=1:detect_leaks=0:symbolize=1:detect_odr_violation=0:print_stacktrace=1" 
But when I use afl-fuzz to run it will report an error
```
ASAN_OPTIONS="disable_coredump=0:unmap_shadow_on_exit=1:abort_on_error=1:detect_leaks=0:symbolize=1:detect_odr_violation=0:print_stacktrace=1" AFL_DEBUG=1 afl-fuzz -i in -o out -m none -- ./1
[+] Enabled environment variable AFL_DEBUG with value 1
[+] Enabled environment variable AFL_DEBUG with value 1
afl-fuzz++4.06a based on afl by Michal Zalewski and a large online community
[+] afl++ is maintained by Marc "van Hauser" Heuse, Heiko "hexcoder" Eißfeldt, Andrea Fioraldi and Dominik Maier
[+] afl++ is open source, get it at https://github.com/AFLplusplus/AFLplusplus
[+] NOTE: This is v3.x which changes defaults and behaviours - see README.md
[+] Enabled environment variable ASAN_OPTIONS with value disable_coredump=0:unmap_shadow_on_exit=1:abort_on_error=1:detect_leaks=0:symbolize=1:detect_odr_violation=0:print_stacktrace=1
[+] No -M/-S set, autoconfiguring for "-S default"
[*] Getting to work...
[+] Using exponential power schedule (FAST)
[+] Enabled testcache with 50 MB
[+] Generating fuzz data with a length of min=1 max=1048576
[*] Checking core_pattern...
[!] WARNING: Could not check CPU scaling governor
[+] You have 4 CPU cores and 6 runnable tasks (utilization: 150%).
[*] Setting up output directories...
[+] Output directory exists but deemed OK to reuse.
[*] Deleting old session data...
[+] Output dir cleanup successful.
[*] Checking CPU core loadout...
[+] Found a free CPU core, try binding to #0.
[*] Scanning 'in'...
[+] Loaded a total of 1 seeds.
[*] Creating hard links for all input files...
[*] Validating target binary...
[*] Spinning up the fork server...
[+] All right - fork server is up.
[*] Extended forkserver functions received (00000000).
[*] No auto-generated dictionary tokens to reuse.
[*] Attempting dry run with 'id:000000,time:0,execs:0,orig:1'...
[D] DEBUG: calibration stage 1/7
string is: Hello world!

=================================================================
==2816702==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 100 byte(s) in 1 object(s) allocated from:
    #0 0x564be3462218 in malloc (/home/ubuntu/1+0xf3218)
    #1 0x564be33774c9 in main /home/ubuntu/1.c:6
    #2 0x7f7ae6c67082 in __libc_start_main ../csu/libc-start.c:308
    #3 0x564be337758d in _start (/home/ubuntu/1+0x858d)

SUMMARY: AddressSanitizer: 100 byte(s) leaked in 1 allocation(s).

[-] Oops, the program crashed with one of the test cases provided. There are
    several possible explanations:

    - The test case causes known crashes under normal working conditions. If
      so, please remove it. The fuzzer should be seeded with interesting
      inputs - but not ones that cause an outright crash.

    - In QEMU persistent mode the selected address(es) for the loop are not
      properly cleaning up variables and memory. Try adding
      AFL_QEMU_PERSISTENT_GPR=1 or select better addresses in the binary.

    - Least likely, there is a horrible bug in the fuzzer. If other options
      fail, poke <afl-users@googlegroups.com> for troubleshooting tips.
[!] WARNING: Test case 'id:000000,time:0,execs:0,orig:1' results in a crash, skipping
[+] All test cases processed.

[-] PROGRAM ABORT : We need at least one valid input seed that does not crash!
         Location : main(), src/afl-fuzz.c:2262
```

**Comment by vanhauser-thc** [comment]
What happens if you do not set ASAN_OPTIONS?

**Comment by cnitlrt** [comment]
thanks for the reply！when I doesn't set the ASAN_OPTIONS,the same error as above will appear.
Maybe the settings doesn’t work?
```
➜  ~ AFL_DEBUG=1 afl-fuzz -i in -o out -m none -- ./1
[+] Enabled environment variable AFL_DEBUG with value 1
[+] Enabled environment variable AFL_DEBUG with value 1
afl-fuzz++4.06a based on afl by Michal Zalewski and a large online community
[+] afl++ is maintained by Marc "van Hauser" Heuse, Heiko "hexcoder" Eißfeldt, Andrea Fioraldi and Dominik Maier
[+] afl++ is open source, get it at https://github.com/AFLplusplus/AFLplusplus
[+] NOTE: This is v3.x which changes defaults and behaviours - see README.md
[+] No -M/-S set, autoconfiguring for "-S default"
[*] Getting to work...
[+] Using exponential power schedule (FAST)
[+] Enabled testcache with 50 MB
[+] Generating fuzz data with a length of min=1 max=1048576
[*] Checking core_pattern...
[!] WARNING: Could not check CPU scaling governor
[+] You have 4 CPU cores and 4 runnable tasks (utilization: 100%).
[*] Setting up output directories...
[+] Output directory exists but deemed OK to reuse.
[*] Deleting old session data...
[+] Output dir cleanup successful.
[*] Checking CPU core loadout...
[+] Found a free CPU core, try binding to #0.
[*] Scanning 'in'...
[+] Loaded a total of 1 seeds.
[*] Creating hard links for all input files...
[*] Validating target binary...
[*] Spinning up the fork server...
[+] All right - fork server is up.
[*] Extended forkserver functions received (00000000).
[*] No auto-generated dictionary tokens to reuse.
[*] Attempting dry run with 'id:000000,time:0,execs:0,orig:1'...
[D] DEBUG: calibration stage 1/7
string is: Hello world!

=================================================================
==2845745==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 100 byte(s) in 1 object(s) allocated from:
    #0 0x55ef8ecd2218  (/home/ubuntu/1+0xf3218)
    #1 0x55ef8ebe74c9  (/home/ubuntu/1+0x84c9)
    #2 0x7fbfc6c08082  (/lib/x86_64-linux-gnu/libc.so.6+0x24082)
    #3 0x55ef8ebe758d  (/home/ubuntu/1+0x858d)

SUMMARY: AddressSanitizer: 100 byte(s) leaked in 1 allocation(s).

[-] Oops, the program crashed with one of the test cases provided. There are
    several possible explanations:

    - The test case causes known crashes under normal working conditions. If
      so, please remove it. The fuzzer should be seeded with interesting
      inputs - but not ones that cause an outright crash.

    - In QEMU persistent mode the selected address(es) for the loop are not
      properly cleaning up variables and memory. Try adding
      AFL_QEMU_PERSISTENT_GPR=1 or select better addresses in the binary.

    - Least likely, there is a horrible bug in the fuzzer. If other options
      fail, poke <afl-users@googlegroups.com> for troubleshooting tips.
[!] WARNING: Test case 'id:000000,time:0,execs:0,orig:1' results in a crash, skipping
[+] All test cases processed.

[-] PROGRAM ABORT : We need at least one valid input seed that does not crash!
         Location : main(), src/afl-fuzz.c:2262
```

**Comment by vanhauser-thc** [comment]
I cannot reproduce this:
```
# cat > x.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(int argc, const char *argv[]) {
    char *s = (char*)malloc(100);
    strcpy(s, "Hello world!");
    printf("string is: %s\n", s);
    return 0;
}

# afl-gcc x.c -o x -fsanitize=address -g -static-libasan
afl-cc++4.07a by Michal Zalewski, Laszlo Szekeres, Marc Heuse - mode: GCC-GCC
[!] WARNING: You are using outdated instrumentation, install LLVM and/or gcc-plugin and use afl-clang-fast/afl-clang-lto/afl-gcc-fast instead!
afl-as++4.07a by Michal Zalewski
[+] Instrumented 3 locations (64-bit, non-hardened mode, ratio 100%).

# afl-fuzz -i in -o out -- ./x
afl-fuzz++4.07a based on afl by Michal Zalewski and a large online community
[+] AFL++ is maintained by Marc "van Hauser" Heuse, Dominik Maier, Andrea Fioraldi and Heiko "hexcoder" Eißfeldt
[+] AFL++ is open source, get it at https://github.com/AFLplusplus/AFLplusplus
[+] NOTE: AFL++ >= v3 has changed defaults and behaviours - see README.md
[+] No -M/-S set, autoconfiguring for "-S default"
[*] Getting to work...
[+] Using exponential power schedule (FAST)
[+] Enabled testcache with 50 MB
[+] Generating fuzz data with a length of min=1 max=1048576
[*] Checking core_pattern...
[*] Checking CPU scaling governor...
[+] You have 8 CPU cores and 2 runnable tasks (utilization: 25%).
[+] Try parallel jobs - see /usr/local/share/doc/afl/fuzzing_in_depth.md#c-using-multiple-cores
[*] Setting up output directories...
[+] Output directory exists but deemed OK to reuse.
[*] Deleting old session data...
[+] Output dir cleanup successful.
[*] Checking CPU core loadout...
[+] Found a free CPU core, try binding to #0.
[*] Scanning 'in'...
[+] Loaded a total of 1 seeds.
[*] Creating hard links for all input files...
[*] Validating target binary...
[*] Spinning up the fork server...
[+] All right - fork server is up.
[*] No auto-generated dictionary tokens to reuse.
[*] Attempting dry run with 'id:000000,time:0,execs:0,orig:in'...
    len = 2, map size = 3, exec speed = 3648 us, hash = fc0e925d8bbe339f
[+] All test cases processed.
[+] Here are some useful stats:

    Test case count : 1 favored, 0 variable, 0 ignored, 1 total
       Bitmap range : 3 to 3 bits (average: 3.00 bits)
        Exec timing : 3648 to 3648 us (average: 3648 us)

[*] No -t option specified, so I'll use an exec timeout of 20 ms.
[+] All set and ready to roll!
```
also with ASAN_OPTIONS - same thing.

BUT - you seem to use an old version of AFL++. can you remove your current installation and compile+install from the stable or dev branch?

**Comment by secnotes** [comment]
I met the same problem with `afl-fuzz++4.09a`.

**Comment by cnitlrt** [comment]
Use docker can solve this problem, but some machines  do not support docker

**Comment by secnotes** [comment]
> Use docker can solve this problem, but some machines do not support docker

You said it. This is a strange problem and only appears in AFL++. Original AFL does not have sach issue.

**Comment by hexcoder-** [comment]
On my older Ubuntu 18.04 (with gcc 7.5.0 and clang 12.0.0) I can reproduce this issue with current dev.

Running the debug build of afl-fuzz, I can see the cause for the termination of afl-fuzz. It is a LeakSanitizer report turned into an error, despite the fact that ASAN_OPTIONS contains `detect_leaks=0`. I have modified the test program to echo the value of the env. variable `ASAN_OPTIONS`.

```
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(int argc, const char *argv[]) {
    char *s = (char*)malloc(100);
    strcpy(s, "Hello world!");
    printf("string is: %s\n", s);
    printf("ASAN_OPTIONS: %s\n", getenv("ASAN_OPTIONS"));
    return 0;
}
```

The output is:
```
$ AFL_DEBUG=1 ASAN_OPTIONS=detect_leaks=0:abort_on_error=1:symbolize=0 ./afl-fuzz -i in -o out -m none -- ./1
[+] Enabled environment variable AFL_DEBUG with value 1
[+] Enabled environment variable AFL_DEBUG with value 1
[+] Enabled environment variable AFL_TMPDIR with value /run/user/1000
afl-fuzz++4.09a based on afl by Michal Zalewski and a large online community
[+] AFL++ is maintained by Marc "van Hauser" Heuse, Dominik Maier, Andrea Fioraldi and Heiko "hexcoder" Eißfeldt
[+] AFL++ is open source, get it at https://github.com/AFLplusplus/AFLplusplus
[+] NOTE: AFL++ >= v3 has changed defaults and behaviours - see README.md
[+] Enabled environment variable ASAN_OPTIONS with value detect_leaks=0:abort_on_error=1:symbolize=0
[+] No -M/-S set, autoconfiguring for "-S default"
[*] Getting to work...
[+] Using exponential power schedule (FAST)
[+] Enabled testcache with 50 MB
[+] Generating fuzz data with a length of min=1 max=1048576
[*] Checking core_pattern...
[*] Checking CPU scaling governor...
[+] You have 32 CPU cores and 1 runnable tasks (utilization: 3%).
[+] Try parallel jobs - see /usr/local/share/doc/afl/fuzzing_in_depth.md#c-using-multiple-cores
[*] Setting up output directories...
[+] Output directory exists but deemed OK to reuse.
[*] Deleting old session data...
[+] Output dir cleanup successful.
[*] Checking CPU core loadout...
[+] Found a free CPU core, try binding to #0.
[*] Scanning 'in'...
[+] Loaded a total of 1 seeds.
[*] Creating hard links for all input files...
[*] Validating target binary...
[*] Spinning up the fork server...
[+] All right - fork server is up.
[*] Extended forkserver functions received (00000000).
[*] No auto-generated dictionary tokens to reuse.
[*] Attempting dry run with 'id:000000,time:0,execs:0,orig:0'...
[D] DEBUG: calibration stage 1/7
string is: Hello world!
ASAN_OPTIONS: detect_leaks=0:abort_on_error=1:symbolize=0

=================================================================
==52032==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 100 byte(s) in 1 object(s) allocated from:
    #0 0x555555619990  (/home/heiko/AFLplusplus/1+0xc5990)
    #1 0x55555555bad5  (/home/heiko/AFLplusplus/1+0x7ad5)
    #2 0x7ffff6e22c86  (/lib/x86_64-linux-gnu/libc.so.6+0x21c86)
    #3 0x55555555bbb9  (/home/heiko/AFLplusplus/1+0x7bb9)

SUMMARY: AddressSanitizer: 100 byte(s) leaked in 1 allocation(s).

[-] Oops, the program crashed with one of the test cases provided. There are
    several possible explanations:

    - The test case causes known crashes under normal working conditions. If
      so, please remove it. The fuzzer should be seeded with interesting
      inputs - but not ones that cause an outright crash.

    - In QEMU persistent mode the selected address(es) for the loop are not
      properly cleaning up variables and memory. Try adding
      AFL_QEMU_PERSISTENT_GPR=1 or select better addresses in the binary.

    - Least likely, there is a horrible bug in the fuzzer. If other options
      fail, poke the Awesome Fuzzing Discord for troubleshooting tips.
[!] WARNING: Test case 'id:000000,time:0,execs:0,orig:0' results in a crash, skipping
[+] All test cases processed.

[-] PROGRAM ABORT : We need at least one valid input seed that does not crash!
         Location : main(), src/afl-fuzz.c:2426
```
The behavior is the same as if there were no ASAN_OPTIONS set.

Funnily, running the program without afl-fuzz works like expected:
```
$ ASAN_OPTIONS=detect_leaks=0:abort_on_error=1:symbolize=0 ./1
string is: Hello world!
ASAN_OPTIONS: detect_leaks=0:abort_on_error=1:symbolize=0
```
with an exit code indicating no error:
```
$ echo $?
0
```

But, if you replace ASAN_OPTIONS with LSAN_OPTIONS then it is working!!
`$ LSAN_OPTIONS=detect_leaks=0:abort_on_error=1:symbolize=0 ./afl-fuzz -i in -o out -m none -- ./1
`

**Comment by hexcoder-** [comment]
@vanhauser-thc maybe we should also set LSAN_OPTIONS in afl-fuzz to be sure...

**Comment by hexcoder-** [comment]
I added another line to the test program for printing the current value of `LSAN_OPTIONS`, which gave:
`LSAN_OPTIONS: exitcode=23:fast_unwind_on_malloc=0:print_suppressions=0:detect_leaks=1:malloc_context_size=30:`

So we to set `LSAN_OPTIONS` with `detect_leaks=1` here (afl-common.c, line 106).

The fix in afl-fuzz would then be to set `LSAN_OPTIONS=detect_leaks=0`, if ASAN_OPTIONS contains `detect_leaks=0` and LSAN_OPTIONS is not set, I think.

**Comment by hexcoder-** [comment]
should be fixed now with commit [6ed3f4c](https://github.com/AFLplusplus/AFLplusplus/commit/6ed3f4cfac52d2bef60321bccb236c7c5d215404).

**Comment by vanhauser-thc** [comment]
hmm that also introduces a new issue:
```
    if (have_asan_options && NULL != strstr(have_asan_options, "detect_leaks=0")
      strcat(buf, "exitcode=" STRINGIFY(LSAN_ERROR) ":fast_unwind_on_malloc=0:pr
    } else {
      should_detect_leaks = 1;
      strcat(buf, "exitcode=" STRINGIFY(LSAN_ERROR) ":fast_unwind_on_malloc=0:pr
    }
```
if NO asan_options are set, then leak sanitizer is activated, also lsan was not set either. this is not what a user wants. I fixed it like:
```
    if (have_asan_options) {

      if (NULL != strstr(have_asan_options, "detect_leaks=0")) {

        strcat(buf, "exitcode=" STRINGIFY(LSAN_ERROR) ":fast_unwind_on_malloc=0
:print_suppressions=0:detect_leaks=0:malloc_context_size=0:");

      } else {

        should_detect_leaks = 1;
        strcat(buf, "exitcode=" STRINGIFY(LSAN_ERROR) ":fast_unwind_on_malloc=0:print_suppressions=0:detect_leaks=1:malloc_context_size=30:");

      }
```

right?
